### PR TITLE
MLDB-1948 levenshtein dist alternate implementation

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: sql/builtin_functions.cc [hunk 1]
```diff
@@ -2495,6 +2495,79 @@ BoundFunction length(const std::vector<BoundSqlExpression> & args)
...
+    else if ((c & 0xF8) == 0xF0) {
+        remaining = 3;
+    }
+    else {
+        throw std::runtime_error(
+            "Invalid UTF-8 character sequence encountered in buffer at index "
+            + to_string(idx));
+    }
+
+    char32_t res = static_cast<char32_t>(c << 8 * remaining);
```
Comment (simlmx) [inline sql/builtin_functions.cc]: Je pense que tu veux tasser ça

File: sql/builtin_functions.cc [hunk 2]
```diff
@@ -2495,6 +2495,79 @@ BoundFunction length(const std::vector<BoundSqlExpression> & args)
...
+    for (; remaining > 0; --remaining) {
+        ++idx;
+        c = in[idx];
+        if ((c & 0xC0) != 0x80) {
+            throw std::runtime_error(
+                "Invalid UTF-8 character sequence encountered in buffer at index "
+                + to_string(idx));
+        }
+        res += static_cast<char32_t>(c << 8 * remaining);
```
Comment (simlmx) [inline sql/builtin_functions.cc]: Pourquoi pas "|=", ou ça revient au même?
Et c'est pas " c << 8 * (remaining-1)"

**Comment by mailletf** [comment]
j'imagine que le PR devrait aussi inclure le retrait de edlib comme sous-module et dans le make system?

File: testing/MLDB-1667_levenshtein_dist.py [hunk 3]
```diff
@@ -53,7 +54,11 @@ def doTestWords(a, b, score):
             "episodes.")
         mldb.log(len(text))
         doTestWords(text, text2, 10)
-        
+
+        doTestWords('Québec', 'Québec', 0)
+        doTestWords('Québec', 'Quebec', 1)
+        doTestWords('éèà', 'abc', 3)
+
```
Comment (simlmx) [inline testing/MLDB-1667_levenshtein_dist.py]: Je rajouterais plus de tests, pcq c'est quand même très possible qu'il y ait une faute dans les math mais qu'on tombe pas dans les edge cases. Je ferais au moins un test avec un utf-8 de chaque longueur possible.

File: sql/builtin_functions.cc (hunk 2 above)
Comment (FinchPowers) [inline sql/builtin_functions.cc]: Tu as raison pour le -1.
Pour le |= je ne connaisais pas. Ça revient au même pour la réponse mais pê pas pour l'opération cpu.

File: sql/builtin_functions.cc [hunk 4]
```diff
@@ -2495,6 +2495,79 @@ BoundFunction length(const std::vector<BoundSqlExpression> & args)
 
 static RegisterBuiltin registerLength(length, "length");
 
+static char32_t getChar32(const string & in, int & idx) {
```
Comment (EricRobert) [inline sql/builtin_functions.cc]: That utf8 decoding code looks buggy to me... I wrote a version a while back and it's a bit more complicated then that. For starter, it can have up to 6 bytes if I remember correctly. See:
https://github.com/EricRobert/rig-string/blob/master/include/utf8.hpp#L123

Comment (FinchPowers) [inline sql/builtin_functions.cc]: As discussed, I implemented RFC 3629 version.

**Comment by FinchPowers** [comment]
We will keep edlib as it is twice faster than this version.
